### PR TITLE
Hash password

### PR DESCRIPTION
**Pull request by rhin0cer0s**
I open this PR regarding the server password storage.
* I used [bcryptjs](https://www.npmjs.com/package/bcryptjs) module. I have genuinely no clue if this
module is secure. I don't have enough nodeJS knowledge to be able to understand it plainly. I choose it mainly because it is full JS and got no dependency.
* I added ```bcrypjs``` dependency to ```package.json```.
* I tried to configure correctly ```tld```. It was the first time so I might miss something. I used a command line close to ```tld install bcryptjs --save```

---
**How it works**
* config.ts - after loading ```config.json``` check if the ```password``` field is a bcrypt string. If not create a bcrypt string corresponding to the ```password``` value, replace it and overwrite the ```config.json``` file. 
* authenticate.ts - use the ```bcrypt.compare``` function to compare the stored value and the received one.

---
**Possible improvements :**
* ```bcryptjs``` module importation is currently really nasty : 
```
/* tslint:disable */
let bcrypt = require("bcryptjs");                                                
/* tslin:enable */
```
  I was not able to make it work with the ```import``` syntax - once again my lack of nodeJS knowledge :(
* merge ```let bcryptRegex = /^\$2a\$10\$/;``` and ```config.password.length !== 60```. My regex skill is a little rusty.

### CODE REVIEW

File: server/config.ts [hunk 1]
```diff
@@ -6,6 +6,11 @@ import configDefaults from "./configDefaults";
 
 export default configDefaults;
 
+/* tslint:disable */
+let bcrypt = require("bcryptjs");
+/* tslin:enable */
+let bcryptRegex = /^\$2a\$10\$/;
```
Comment (elisee) [inline server/config.ts]: I think the following should do it: `/^\$2a\$10\$.{54}/`, [interpreted by Regexper](http://regexper.com/#%2F%5E%5C%242a%5C%2410%5C%24.%7B54%7D%2F).

File: server/authenticate.ts [hunk 2]
```diff
@@ -10,13 +14,38 @@ export default function(socket: SocketIO.Socket, next: Function) {
...
+    next(new Error("invalidUsername"));
+    return;
   }
-  next();
+
+  verifyServerPassword(auth.serverPassword, config.password, socket, next);
+  return;
+}
+
+function verifyServerPassword(password: string, hash: string, socket: SocketIO.Socket, next : Function) {
```
Comment (elisee) [inline server/authenticate.ts]: There's an extra space before the colo in `next : Function`

File: server/authenticate.ts [hunk 3]
```diff
@@ -1,5 +1,9 @@
 import config from "./config";
-
+
+/* tslint:disable */
+let bcrypt = require("bcryptjs");
```
Comment (elisee) [inline server/authenticate.ts]: `import * as bcrypt from "bcryptjs";` should do the trick.

**Comment by elisee** [comment]
Thanks for looking into this. Added a few comments, and two things I'm wondering about, although I haven't tested the PR yet:

First, I think the launcher needs to be updated to deal with the fact that the stored password is now a bcrypt hash. It should probably not display 60 ●●● characters? How should it behave?

Second, unless I'm wrong, this fixes storing the password as plain text on the server side, but the client-side cookie will still be storing it in plain text. I don't think this can be dealt with without setting up a proper session system as described in #55.

Also, When we switch to the new authentication system, do we still want to have a user-definable server password at all? Maybe we should just allow the server owner to share a secret signup link: `http://1.2.3.4:4237/signup?secret=LONG_RANDOM_STRING` with an option to invalidate the current secret and generate a new one? In which case, we won't need to use bcrypt to store that secret (but we would still use it for storing local user passwords!)

What do you think?

File: server/config.ts (hunk 1 above)
Comment (rhin0cer0s) [inline server/config.ts]: A bcrypt string is 60 characters long. ```$2a$10$``` is 7 long. So I put ```/^\$2a\$10\$.{53}$/```. It seems to work.

File: server/authenticate.ts (hunk 2 above)
Comment (rhin0cer0s) [inline server/authenticate.ts]: I removed it, also removed some space in the anonymous function called during ```bcrypt.compare```.

File: server/authenticate.ts (hunk 3 above)
Comment (rhin0cer0s) [inline server/authenticate.ts]: Yeah it works... I don't understand why at all. When I tried the first time it definitely did not...

**Comment by rhin0cer0s** [comment]
> First, I think the launcher needs to be updated to deal with the fact that the stored password is now a bcrypt hash. It should probably not display 60 ●●● characters? How should it behave?

Aw. Since I use SP with command line / browser I did not see this problem. It should behave differently. Furthermore the server **must be** reboot in order to generate the bcrypt string. We should also look into this.

> Second, unless I'm wrong, this fixes storing the password as plain text on the server side, but the client-side cookie will still be storing it in plain text. I don't think this can be dealt with without setting up a proper session system as described in #55.

Yup. I don't know how the cookie auth works but storing a password - plain text or hashed - is really wrong. I hope the use of ```passport``` correct this point.

> Also, When we switch to the new authentication system, do we still want to have a user-definable server password at all? Maybe we should just allow the server owner to share a secret signup link: http://1.2.3.4:4237/signup?secret=LONG_RANDOM_STRING with an option to invalidate the current secret and generate a new one? In which case, we won't need to use bcrypt to store that secret (but we would still use it for storing local user passwords!)

The temporary - or at least "private" - link might be a nice solution. About storing the user password it depends on the authentication process. Since you wish to use ```passport``` I think it exists a module which can handle connection process without the use of an outside service. ```passport-local``` seems to be the one. But I don't know how it stores its password/login duple.

I know nothing is fixed about the authentication process. But ```passport``` will handle everything just fine I think.

**Comment by elisee** [comment]
> Yup. I don't know how the cookie auth works but storing a password - plain text or hashed - is really wrong. I hope the use of passport correct this point.

Yes, `passport` has a proper session system. To be clear: the plain text password was always just a stopgap measure and never meant to last. The plan was to use `bcrypt` and a proper session system all along. It only lingered because, well, life happens.

> `passport-local` seems to be the one. But I don't know how it stores its password/login duple.

Passport lets one defines how things are stored so we should totally plug `bcrypt` behind it.

**Comment by elisee** [comment]
Closing this PR for now since we're not going to merge it as is. Let's track progress on better auth in #55, with the explicit goal being to have it done by the next release or, at worst, the one after.

**Pull request closed without merging**